### ISSUE TITLE
mismatched fingerprints for core/v2.3.3

### ISSUE THREAD

**Issue by syzygyzer** [trezor/trezor-firmware#1256]
**Describe the bug**
Possibly related to https://github.com/trezor/trezor-firmware/issues/692

I am attempting to build the latest Trezor Model T firmware following the instructions at https://wiki.trezor.io/Developers_guide:Deterministic_firmware_build, except I checkout core/2.3.3 instead of core/2.3.0 it currently states.

I am running a fairly fresh minimal install of Debian Stable 10 on x64. I have userspace managed mostly with Nix, but this is the first Python build I've done on this system so far and installed the prerequisites up through `wheel` listed here globally https://wiki.trezor.io/Using_trezorctl_commands_with_Trezor using pip3. But theoretically my system's environment shouldn't matter inside of a Docker -> Nix -> `pipenv` triple combo! This is also the first build I've done with `docker` on this system, so all I have installed is the most recent stable version.

I've tried the build a few with minor variations a few times and the output fingerprints have always been the same. I followed it as written the first time. I noticed a message saying `LANG` was unset and the compiler suggested to set it in `.profile`, so I did even though `echo $LANG` gave `en.US-UTF8`  in the shell (for which I tried both `fish` and `bash`). Each time I got some intermediate fingerprint messages that look like this:

```
Vendor Header for SatoshiLabs version 0.0 (4608 bytes)                                                                                                                                                                                       
✔ Signature is VALID                                                                                                                                                                                                                         
Firmware Header {                                                                                                                                                                                                                            
    magic: HeaderType.FIRMWARE                                                                                                                                                                                                               
    header_len: 1024                                                                                                                                                                                                                         
    expiry: 0                                                                                                                                                                                                                                
    code_length: 1079296                                                                                                                                                                                                                     
    version: 2.3.3 build 0                                                                                                                                                                                                                   
    fix_version: 2.2.0 build 0                                                                                                                                                                                                               
    hashes: [                                                                                                                                                                                                                                
        ❌ 0000000000000000000000000000000000000000000000000000000000000000                                                                                                                                                                  
        ❌ 0000000000000000000000000000000000000000000000000000000000000000                                                                                                                                                                  
        ❌ 0000000000000000000000000000000000000000000000000000000000000000                                                                                                                                                                  
        ❌ 0000000000000000000000000000000000000000000000000000000000000000                                                                                                                                                                  
        ❌ 0000000000000000000000000000000000000000000000000000000000000000                                                                                                                                                                  
        ❌ 0000000000000000000000000000000000000000000000000000000000000000                                                                                                                                                                  
        ❌ 0000000000000000000000000000000000000000000000000000000000000000                                                                                                                                                                  
        ❌ 0000000000000000000000000000000000000000000000000000000000000000                                                                                                                                                                  
        ❌ 0000000000000000000000000000000000000000000000000000000000000000                                                                                                                                                                  
        ✔ 0000000000000000000000000000000000000000000000000000000000000000                                                                                                                                                                   
        ✔ 0000000000000000000000000000000000000000000000000000000000000000                                                                                                                                                                   
        ✔ 0000000000000000000000000000000000000000000000000000000000000000                                                                                                                                                                   
        ✔ 0000000000000000000000000000000000000000000000000000000000000000                                                                                                                                                                   
        ✔ 0000000000000000000000000000000000000000000000000000000000000000                                                                                                                                                                   
        ✔ 0000000000000000000000000000000000000000000000000000000000000000                                                                                                                                                                   
        ✔ 0000000000000000000000000000000000000000000000000000000000000000                                                                                                                                                                   
    ]                                                                                                                                                                                                                                        
    sigmask: 0                                                                                                                                                                                                                               
    signature: 64 bytes 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000                                                                                     
}                                                                                                                                                                                                                                            
Fingerprint: dda77cd7893a5f413f8fc4b2f44d1d43ed4b26e8ced5e6e578cc6b302c1a2310                                                                                                                                                                
❌ Signature is MISSING, hashes are MISSING                                                                                                                                                                                                  
No changes made                                                                                                                                                                                                                              
dd if=build/firmware/firmware.bin of=build/firmware/firmware.bin.p1 skip=0 bs=128k count=6                                                                                                                                                   
6+0 records in                                                                                                                                                                                                                               
6+0 records out                                                                                                                                                                                                                              
786432 bytes (786 kB, 768 KiB) copied, 0.000876145 s, 898 MB/s                                                                                                                                                                               
Cloning into 'trezor-firmware'...
```

where there's a big ole block of zeroes and then the correct fingerprint, and at the end I see printed

```
Fingerprints:
acb3d55cb2358b4c6bfc7b008299b08f694ff7c7ef3c03b8e532f2270779dc30 build/core/firmware/firmware.bin
7b6dc5179c3f273430c230abeb4c0174ef2120ecc783d72d33ddd86e63790430 build/core-bitcoinonly/firmware/firmware.bin
4406a18f8553940dba0ce17409b8c87f8502c2741b8d1a74ad70095dd397a8bf build/legacy/firmware/firmware.bin
27276dc311f253e15f7c10d49520b03ab477912b2119ef441fd6a7777fef1b6c build/legacy-bitcoinonly/firmware/firmware.bin
```

This does not match with the v2.3.3 (or any version) shared here: https://wallet.trezor.io/data/firmware/2/releases.json 

```
"fingerprint": "46326222f8afcb82e1cd07867bc3bf8836f4e9d0f367e23b58d1e9bc32cd032e",
"fingerprint_bitcoinonly": "dda77cd7893a5f413f8fc4b2f44d1d43ed4b26e8ced5e6e578cc6b302c1a2310",
```

Furthermore, checking the fingerprints using the other method on the deterministic page with `tail` gives another result:

```
me@me ~/trezor/trezor-firmware/build/core/firmware$ tail -c +1024 firmware.bin | sha256sum
43af0f10fa77d17a2653cb11825395fe27cd0197125fdd0f7d2981462ad0b607  -
me@me ~/trezor/trezor-firmware/build/core-bitcoinonly/firmware$ tail -c +1024 firmware.bin | sha256sum
cf0d90d67cdd956ff6536011e2cd3fed59b4acc94989793afec06a16ed2e8b1b
```
Trying `tail -c +1280` also does not match anything.

Possibly this is not a bug at all and the fingerprints shown during the build process are the correct ones, and the rest should be ignored. If so you can simply close the issue. But the `❌ Signature is MISSING, hashes are MISSING` seems odd. I have never tried to build the Trezor software before today though.

I also tried to build core/v2.3.0 once but it failed to complete somehow - not sure if this will help diagnose my issue with v2.3.3. I have not tried to build any other versions yet.

**Firmware version and revision**
core/v2.3.3

**Desktop/smartphone setup (please complete the following information):**
 - Wallet software: trezor-firmware
 - OS: Debian Stable 
 - Version - 10

Thank you!

**Comment by prusnak** [comment]
We should make sure the hashes are added to the firmware header first and compute the fingerprint only after this has been done. Can you please have a look at this @mmilata? Maybe @matejcik can help with the tooling.

**Comment by matejcik** [comment]
how are the final fingerprints calculated? fingerprint for core should be calculated by `headertool.py --digest`, which takes care of the missing hashes internally. Or alternately `headertool.py --rehash` should be used to insert the hashes; this usually happens in the same step as signing, as I thought that hashing without signing is kind of pointless.

**Comment by matejcik** [comment]
oooh i see, it's using the `header_digest` function from `trezorlib.firmware`
well, that's not right.

IMHO the behavior of `header_digest` is correct in ignoring the bad hash values; there is no signing capability and if you tried to upload hashless firmware, it would in fact have the digest that `header_digest` claims.

So the solution is to either use `headertool` for generating the `.fingerprint` file, or using `headertool --rehash` before

**Comment by prusnak** [comment]
@matejcik it seems martin used the `trezorlib.firmware.digest` function because it supports both T1 and TT. I think we can use the `firmware-fingerprint.py` tool as is for T1, but use `headertool -d` for TT?

**Comment by matejcik** [comment]
yes, that's reasonable

**Comment by mmilata** [comment]
Thank you for quality bug report, @syzygyzer! Proposed fix is in #1261. 

Regarding the `tail -c ... | sha256sum` method from the Trezor Wiki, that is for Trezor One only. Trezor T currently has 5632 bytes long vendor header + firmware header, so the following commands should produce identical hashes:

```
tail -c +5633 trezor-2.3.3.bin | sha256sum
tail -c +5633 trezor-2.3.3-your-build.bin | sha256sum
```
`tail -c +N` seems to be 1-based and signature is at the very end, thus 5633.
